File: v1/scala/pig-latin/src/main/scala/PigLatin.scala
object PigLatin {
  val vowelRegex = """(?<begin>^|\s+)(?<vowel>a|e|i|o|u|yt|xr)(?<rest>\w+)""".r
  val consonantRegex = """(?<begin>^|\s+)(?<consonant>ch|qu|thr|th|sch|yt|\wqu|\w)(?<rest>\w+)""".r

  val vowelReplacement = "${begin}${vowel}${rest}ay"
  val consonantReplacement = "${begin}${rest}${consonant}ay"

  def translate(sentence: String) =
    if (vowelRegex.findFirstIn(sentence).nonEmpty)
      vowelRegex.replaceAllIn(sentence, vowelReplacement)
  else
      consonantRegex.replaceAllIn(sentence, consonantReplacement)
}

File: v1/csharp/pig-latin/PigLatin.cs
using System.Text.RegularExpressions;

public static class PigLatin
{
    private static readonly Regex VowelRegex = new Regex(@"(?<begin>^|\s+)(?<vowel>a|e|i|o|u|yt|xr)(?<rest>\w+)", RegexOptions.Compiled);
    private static readonly Regex ConsonantRegex = new Regex(@"(?<begin>^|\s+)(?<consonant>ch|qu|thr|th|sch|yt|\wqu|\w)(?<rest>\w+)", RegexOptions.Compiled);

    private const string VowelReplacement = "${begin}${vowel}${rest}ay";
    private const string ConsonantReplacement = "${begin}${rest}${consonant}ay";

    public static string Translate(string sentence)
    {
        if (VowelRegex.IsMatch(sentence))
        {
            return VowelRegex.Replace(sentence, VowelReplacement);
        }

        return ConsonantRegex.Replace(sentence, ConsonantReplacement);
    }
}